src/box/iproto.cc

```diff
@@ -1523,11 +1761,122 @@ tx_process1(struct cmsg *m)
 		goto error;
 	if (tuple && tuple_to_obuf(tuple, out))
 		goto error;
+	fiber_set_txn(fiber(), NULL);
 	iproto_reply_select(out, &svp, msg->header.sync, ::schema_version,
 			    tuple != 0);
 	iproto_wpos_create(&msg->wpos, out);
 	return;
 error:
+	fiber_set_txn(fiber(), NULL);
+	tx_reply_error(msg);
+}
+
+static void
+tx_process_begin(struct cmsg *m)
+{
+	struct iproto_msg *msg = tx_accept_msg(m);
+	struct obuf *out;
+	struct stream *stream = msg->stream;
+
+	if (tx_check_schema(msg->header.schema_version))
+		goto error;
+
+	if (stream == NULL) {
+		diag_set(ClientError, ER_PROTOCOL,
+			 "Missing stream_id for iproto "
+			 "`begin` request");
+		goto error;
+	}
+
+	if (stream->txn != NULL) {
+		diag_set(ClientError, ER_ACTIVE_TRANSACTION);
+		goto error;
+	}
+
+	if (box_txn_begin() != 0)
+		goto error;
+
+	stream->txn = in_txn();
```

Comment:
Oh, that's the thing I wanted to avoid. I fear this juggling looks so ugly that we could leave it so only if we swear to refactor it later.